environment/config.go

```diff
@@ -0,0 +1,221 @@
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	etcdclient "github.com/m3db/m3cluster/client/etcd"
+	"github.com/m3db/m3cluster/kv"
+	m3clusterkvmem "github.com/m3db/m3cluster/kv/mem"
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3cluster/shard"
+	"github.com/m3db/m3db/kvconfig"
+	"github.com/m3db/m3db/retention"
+	"github.com/m3db/m3db/sharding"
+	"github.com/m3db/m3db/storage/namespace"
+	"github.com/m3db/m3db/topology"
+	"github.com/m3db/m3db/ts"
+	"github.com/m3db/m3x/instrument"
+)
+
+var (
+	errNilRetention  = errors.New("namespace retention options cannot be empty")
+	errMissingConfig = errors.New("must supply service or static config")
+)
+
+// Configuration is a configuration that can be used to create namespaces, a topology, and kv store
+type Configuration struct {
+	// Service is used when a topology initializer is not supplied.
+	Service *etcdclient.Configuration `yaml:"service"`
+
+	// StaticConfiguration is used for running M3DB with a static config
+	Static *StaticConfiguration `yaml:"static"`
+}
+
+// StaticConfiguration is used for running M3DB with a static config
+type StaticConfiguration struct {
+	Namespaces     []StaticNamespaceConfiguration `yaml:"namespaces"`
+	TopologyConfig *topology.StaticConfiguration  `yaml:"topology"`
+	ListenAddress  string                         `yaml:"listenAddress"`
+}
+
+// StaticNamespaceConfiguration sets the static namespace
+type StaticNamespaceConfiguration struct {
+	Name      string                    `yaml:"name"`
+	Options   *StaticNamespaceOptions   `yaml:"options"`
+	Retention *StaticNamespaceRetention `yaml:"retention"`
+}
+
+// StaticNamespaceOptions sets namespace options- if nil, default is used
+type StaticNamespaceOptions struct {
+	NeedsBootstrap      bool `yaml:"needsBootstrap"`
+	NeedsFlush          bool `yaml:"needsFlush"`
+	WritesToCommitLog   bool `yaml:"writesToCommitLog"`
+	NeedsFilesetCleanup bool `yaml:"needsFilesetCleanup"`
+	NeedsRepair         bool `yaml:"needsRepair"`
+}
+
+// StaticNamespaceRetention sets the retention per namespace (required)
+type StaticNamespaceRetention struct {
+	RetentionPeriod                     time.Duration `yaml:"retentionPeriod"`
+	BlockSize                           time.Duration `yaml:"blockSize"`
+	BufferFuture                        time.Duration `yaml:"bufferFuture"`
+	BufferPast                          time.Duration `yaml:"bufferPast"`
+	BlockDataExpiry                     bool          `yaml:"blockDataExpiry"`
+	BlockDataExpiryAfterNotAccessPeriod time.Duration `yaml:"blockDataExpiryAfterNotAccessPeriod"`
+}
+
+// ConfigureResults stores initializers and kv store for dynamic and static configs
+type ConfigureResults struct {
+	NamespaceInitializer namespace.Initializer
+	TopologyInitializer  topology.Initializer
+	KVStore              kv.Store
+}
+
+// Configure creates a new ConfigureResults
+func (c Configuration) Configure(iopts instrument.Options, hashingSeed uint32) (ConfigureResults, error) {
```

Comment:
Can we make the argument to this method a struct so if it changes it's not a breaking change to callers in the future?  i.e.